# DATA UNDERSTANDING

## Load and Inspect the data

In [1]:
# Import necessary Libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')

In [2]:
# Loading the dataset
df = pd.read_csv("Electric_Vehicle_Population_Data.csv")

In [3]:
# Check the data head
df.head()

VIN (1-10)     County       City State  Postal Code  Model Year   Make  \
0  5YJYGDEE1L       King    Seattle    WA      98122.0        2020  TESLA   
1  7SAYGDEE9P  Snohomish    Bothell    WA      98021.0        2023  TESLA   
2  5YJSA1E4XK       King    Seattle    WA      98109.0        2019  TESLA   
3  5YJSA1E27G       King   Issaquah    WA      98027.0        2016  TESLA   
4  5YJYGDEE5M     Kitsap  Suquamish    WA      98392.0        2021  TESLA   

     Model           Electric Vehicle Type  \
0  MODEL Y  Battery Electric Vehicle (BEV)   
1  MODEL Y  Battery Electric Vehicle (BEV)   
2  MODEL S  Battery Electric Vehicle (BEV)   
3  MODEL S  Battery Electric Vehicle (BEV)   
4  MODEL Y  Battery Electric Vehicle (BEV)   

   Clean Alternative Fuel Vehicle (CAFV) Eligibility  Electric Range  \
0            Clean Alternative Fuel Vehicle Eligible             291   
1  Eligibility unknown as battery range has not b...               0   
2            Clean Alternative Fuel Vehicle Eligible             270   
3            Clean Alternative Fuel Vehicle Eligible             210   
4  Eligibility unknown as battery range has not b...               0   

   Base MSRP  Legislative District  DOL Vehicle ID  \
0          0                  37.0       125701579   
1          0                   1.0       244285107   
2          0                  36.0       156773144   
3          0                   5.0       165103011   
4          0                  23.0       205138552   

                Vehicle Location  \
0   POINT (-122.30839 47.610365)   
1  POINT (-122.179458 47.802589)   
2   POINT (-122.34848 47.632405)   
3   POINT (-122.03646 47.534065)   
4   POINT (-122.55717 47.733415)   

                                Electric Utility  2020 Census Tract  
0   CITY OF SEATTLE - (WA)|CITY OF TACOMA - (WA)       5.303301e+10  
1                         PUGET SOUND ENERGY INC       5.306105e+10  
2   CITY OF SEATTLE - (WA)|CITY OF TACOMA - (WA)       5.303301e+10  
3  PUGET SOUND ENERGY INC||CITY OF TACOMA - (WA)       5.303303e+10  
4                         PUGET SOUND ENERGY INC       5.303594e+10

Based on the above code output it shows 5 Tesla electric vehicles registered in Washington State, primarily in King County. All are Battery Electric Vehicles (BEVs), with models including the Model Y and Model S.

However, there are inconsistencies in the data:

Electric Range of 0 appears in 3 out of 5 entries, which is likely a garbage or missing value, especially for models like the Model Y or S, which typically have substantial range.

Base MSRP is 0 for all entries, suggesting incomplete or unreported pricing data.

The "CAFV Eligibility" is listed as "unknown" for some vehicles, indicating a lack of battery range data.

This suggests a need for data cleaning, particularly for range and price fields, before any meaningful analysis or policy decisions can be made from this dataset.

In [4]:
# Check the data head
df.tail()

VIN (1-10)  County           City State  Postal Code  Model Year  \
177861  7SAYGDEE3N  Pierce    Bonney Lake    WA      98391.0        2022   
177862  KM8K23AG1P   Mason        Shelton    WA      98584.0        2023   
177863  5YJYGDEE6M   Grant         Quincy    WA      98848.0        2021   
177864  WVGKMPE27M    King  Black Diamond    WA      98010.0        2021   
177865  5YJ3E1EA8M  Pierce         Tacoma    WA      98422.0        2021   

              Make          Model           Electric Vehicle Type  \
177861       TESLA        MODEL Y  Battery Electric Vehicle (BEV)   
177862     HYUNDAI  KONA ELECTRIC  Battery Electric Vehicle (BEV)   
177863       TESLA        MODEL Y  Battery Electric Vehicle (BEV)   
177864  VOLKSWAGEN           ID.4  Battery Electric Vehicle (BEV)   
177865       TESLA        MODEL 3  Battery Electric Vehicle (BEV)   

        Clean Alternative Fuel Vehicle (CAFV) Eligibility  Electric Range  \
177861  Eligibility unknown as battery range has not b...               0   
177862  Eligibility unknown as battery range has not b...               0   
177863  Eligibility unknown as battery range has not b...               0   
177864  Eligibility unknown as battery range has not b...               0   
177865  Eligibility unknown as battery range has not b...               0   

        Base MSRP  Legislative District  DOL Vehicle ID  \
177861          0                  31.0       195224452   
177862          0                  35.0       228454180   
177863          0                  13.0       168797219   
177864          0                   5.0       182448801   
177865          0                  27.0       211464683   

                       Vehicle Location  \
177861     POINT (-122.183805 47.18062)   
177862    POINT (-123.105305 47.211085)   
177863  POINT (-119.8493873 47.2339933)   
177864     POINT (-122.00451 47.312185)   
177865      POINT (-122.38578 47.28971)   

                                         Electric Utility  2020 Census Tract  
177861      PUGET SOUND ENERGY INC||CITY OF TACOMA - (WA)       5.305307e+10  
177862  BONNEVILLE POWER ADMINISTRATION||CITY OF TACOM...       5.304596e+10  
177863                           PUD NO 2 OF GRANT COUNTY       5.302501e+10  
177864      PUGET SOUND ENERGY INC||CITY OF TACOMA - (WA)       5.303303e+10  
177865  BONNEVILLE POWER ADMINISTRATION||CITY OF TACOM...       5.305394e+10

All five electric vehicles listed are recent Battery Electric Vehicles (BEVs) (model years 2021–2023) from brands like Tesla, Hyundai, and Volkswagen, registered across various Washington counties.

However, all entries have missing or zero values for:

Electric Range

Base MSRP

CAFV Eligibility

This strongly indicates incomplete or faulty data entry, making the records unreliable for any analysis involving vehicle performance, eligibility, or pricing. Data validation and enrichment are needed before further use.

In [6]:
# Check the data shape
df.shape

(177866, 17)

The output (177866, 17) indicates the shape of the dataset — specifically:

177,866 rows (individual electric vehicle records)

17 columns (features/attributes per vehicle)

🔍 Quick Insight:
The dataset is large and comprehensive, suggesting it's a valuable resource for analyzing electric vehicle trends in Washington State. However, based on the earlier sample rows, data quality issues like missing values e.g. in electric range, MSRP etc may need to be addressed before conducting accurate analyses.

In [7]:
# Chack the data size
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 177866 entries, 0 to 177865
Data columns (total 17 columns):
 #   Column                                             Non-Null Count   Dtype  
---  ------                                             --------------   -----  
 0   VIN (1-10)                                         177866 non-null  object 
 1   County                                             177861 non-null  object 
 2   City                                               177861 non-null  object 
 3   State                                              177866 non-null  object 
 4   Postal Code                                        177861 non-null  float64
 5   Model Year                                         177866 non-null  int64  
 6   Make                                               177866 non-null  object 
 7   Model                                              177866 non-null  object 
 8   Electric Vehicle Type                              177866 non-null  object

The dataset contains 177,866 electric vehicle records across 17 columns. While most fields are complete, there are some minor but notable gaps:

Missing values in key location fields like:

County, City, Postal Code (5 missing)

Legislative District (~389 missing)

Vehicle Location, Electric Utility, 2020 Census Tract (5–9 missing)

Data types are appropriate:

Categorical/object: For vehicle details (e.g., Make, Model, CAFV eligibility)

Numerical: For year, range, price, district, etc.

⚠️ Key Flag:
Even though Electric Range and Base MSRP have no nulls, earlier samples show values of 0, indicating garbage or placeholder data that should be cleaned despite appearing complete.

In [8]:
#Discriptive statistics
df.describe()

Postal Code     Model Year  Electric Range      Base MSRP  \
count  177861.000000  177866.000000   177866.000000  177866.000000   
mean    98172.453506    2020.515512       58.842162    1073.109363   
std      2442.450668       2.989384       91.981298    8358.624956   
min      1545.000000    1997.000000        0.000000       0.000000   
25%     98052.000000    2019.000000        0.000000       0.000000   
50%     98122.000000    2022.000000        0.000000       0.000000   
75%     98370.000000    2023.000000       75.000000       0.000000   
max     99577.000000    2024.000000      337.000000  845000.000000   

       Legislative District  DOL Vehicle ID  2020 Census Tract  
count         177477.000000    1.778660e+05       1.778610e+05  
mean              29.127481    2.202313e+08       5.297672e+10  
std               14.892169    7.584987e+07       1.578047e+09  
min                1.000000    4.385000e+03       1.001020e+09  
25%               18.000000    1.814743e+08       5.303301e+10  
50%               33.000000    2.282522e+08       5.303303e+10  
75%               42.000000    2.548445e+08       5.305307e+10  
max               49.000000    4.792548e+08       5.603300e+10

In [9]:
#Check the missing values
df.isnull().sum()

VIN (1-10)                                             0
County                                                 5
City                                                   5
State                                                  0
Postal Code                                            5
Model Year                                             0
Make                                                   0
Model                                                  0
Electric Vehicle Type                                  0
Clean Alternative Fuel Vehicle (CAFV) Eligibility      0
Electric Range                                         0
Base MSRP                                              0
Legislative District                                 389
DOL Vehicle ID                                         0
Vehicle Location                                       9
Electric Utility                                       5
2020 Census Tract                                      5
dtype: int64

The summary reveals serious data quality issues in two critical fields:

🔋 Electric Range:

Median = 0 and 75% percentile = 75 → At least half the vehicles have no recorded range, likely indicating missing or unreported values.

💲 Base MSRP:

Mean = $1,073, Median = $0, 75% percentile = $0 → Almost 75% of entries have zero MSRP, which is implausible and signals widespread missing data.

⚠️ Key Takeaway:
Despite numeric completeness, over half the records have placeholder values (0) for essential fields like range and price, making them unreliable for analysis without imputation or filtering.

In [10]:
#Finding the dublicates values
df.duplicated().sum()

np.int64(0)

Based onthe above output there is no duplicate values